openml-api/src/main/java/com/feedzai/openml/explanations/ModelExplainer.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2022 Feedzai
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.feedzai.openml.explanations;
+
+import com.feedzai.openml.data.Instance;
+import com.feedzai.openml.model.MachineLearningModel;
+
+/**
+ * A base class for explanation algorithms. Holds any {@link MachineLearningModel} and can compute the
+ * feature contributes for its predictions through {@link #getFeatureContributions(Instance)}.
+ *
+ * @since 1.1.1 //QUESTION are we supposed to stop using @since in all repos?
+ */
+public interface ModelExplainer<M extends MachineLearningModel> {
```

Comment:
instead of creating a wrapper, I would recommend to create a new interface that would extend `ClassificationMLModel `

Reply:
the idea here is to decoupled from ClassificationMLModel, for my understanding, we might have explainers for RegrassionMLModel.

Reply:
yeah but this forces you to create this wrapper when we want to get explanations

while if we create two new interfaces one for ClassificationMLModel and other for RegrassionMLModel(?), then we could get the explanation at the same level where we are calling the method to score instances

Reply:
Yep, as always, every solution has a few trade-offs and that is one downside of this one. 
What you are suggesting was our initial approach, but after several iterations me and Sheng have converged to the current one, because we believe the benefits far outweigh such disadvantages. The advantages we thought of are:

- As Sheng mentioned, we never know what the future might bring us. We currently have ClassificationModel, RegressionModel, but in the future  we might have OutlierDetectionModel, ReinforcementLearningModel, etc, who knows.  In that scenario, the current approach would lead to a far less complex codebase.
- There is more isolation of responsibility. We have the possibility to separate the explanation logic from the model implementation class.
- Each model doesn't need to be tightly coupled with one explanation algorithm. Allowing us to change between several explanation algorithms for the same model (which I am confident that will eventually be necessary, even if far from now).
- It will be easier to implement model agnostic explanation algorithms, which [is a thing](https://homes.cs.washington.edu/~marcotcr/blog/lime/).
 
With that said, we believe this approach is far more flexible and future-proof. Regarding your point, I think we have done the [implementation in Pulse](https://review.feedzai.com/D33573) in a way that compensates for it, but please take a look if you have the time.

We will be happy to discuss offline if you want

Reply:
I forgot to mention one of the points, which was actually what initially lead us to this approach.
Having to implement the explanations interface within the model might require making a lot of changes to it. E.g., the FZRF explanations require that a parameter `classIndex` was provided.
That caused us to have to make several changes in the model, its config class, its factory class, just because of the explanations. This way it is completely separated.